### ISSUE TITLE
Suggestion: Upper-bound generic type constraints

### ISSUE THREAD

**Issue by GregRos** [Microsoft/TypeScript#9252]
This is a proposal for generic structural supertype constraints.

I'll first detail the proposal, and then get into a discussion about use-cases.
## Proposal

A supertype constraint is a constraint on a generic type parameter to be the structural _supertype_ of another type. It's like a subtype/inheritance constraint, but from the other direction. For example, let's say we have a square:

```
interface Square {width : number; height : number}
```

In this suggested syntax, you could write:

```
performAction<T extended by Square>(partialSquare : T) {}
```

This means the type `T` must be a structural supertype of square. So the potential candidates for `T` are:

```
{width : number}, {height : number}, Square, {}
```

It's an established kind of type constraint, not something I just made up. Although it's not exactly common, some languages do implement this feature, such as Scala. In Scala, you can write:

```
def performAction[T >: Square](obj : T) = { ... }
```

To express a supertype/upper bound constraint. In this case, of course, the constraint isn't structural -- `T` must declare that it implements `Square`. 
## Utility

In most languages, including Scala, this kind of constraint isn't very useful. It only comes up in certain specific situations. 

However, Javascript libraries often have this kind of API, where you're allowed to specify the partial properties of an object to modify it (the rest remain at their previous value).

In many cases, you can support this by having optional interface members, but isn't always possible or correct.

An important example is React, which defines a method called `setState`:

```
class Component {
    var state;

    setState(partialState) {
        //merges the properties of partialState with the current state.
    }
}
```

The right signature for this method should be:

```
class Component<Props, State> {
    setState<PartialState extended by State>(partialState : PartialState) {
        //merges the properties of partialState with the current state.
    }
}
```

Currently,the definition files state that it is:

```
setState(fullState : State) {
    //merges the properties of partialState with the current state.
}
```

Which doesn't fully capture the functionality of the method.
## Notes

The suggested syntax doesn't give us a nice way of combining both subtype and supertype constraints. One possibility is:

```
exampleMethod<T extends LowerBound and extended by UpperBound>
```

**Comment by Strate** [comment]
Possible duplicate of
#4889
#371

**Comment by GregRos** [comment]
@Strate, Those really aren't duplicates at all. They're very different, both conceptually and practically.

That said, I'm quite surprised this hasn't been proposed previously, so I suspected it was a duplicate, although I couldn't find anything.

**Comment by Strate** [comment]
@GregRos that tickets (especially #4889) contains discussion very close to your suggestion. For example, see [this comment](https://github.com/Microsoft/TypeScript/issues/4889#issuecomment-200388292), or [that](https://github.com/Microsoft/TypeScript/issues/4889#issuecomment-200745084). I wouldn't say that is it *very different*.

**Comment by GregRos** [comment]
@State yup, those comments are exactly my proposal, though the proposals themselves are quite different. I'm not sure if that means this is a duplicate or not.

**Comment by GregRos** [comment]
@State yup, those comments are exactly my proposal, though the proposals themselves are different. I'm not sure if that means this is a duplicate or not.

Incidentally, I had no idea Java supported this kind of constraint. I thought it was just a Scala thing.

On second thought, #4889 can be used to get the same effect in some cases. One example where it can't is the following:

    blah<Partial extended by Full>(partial : Partial) : Partial {
        return partial;
    } //type information is preserved.

And one example where that proposal can do something that mine can't:

    blah() : partial Full { //Full : {a : number, b : number}
        return {}; //or:
        return {a : 4}; //etc
    }

e.g. The function can return different supertypes of `Full`. This partial type thing is basically the same as a supertype-constrained existential type. It lets you say "X is some supertype of Foo, but no one knows which one".

#371 is similar to #4889. So you're right, on second thought, they're not *very different*.

**Comment by pauldraper** [comment]
> In most languages, including Scala, this kind of constraint isn't very useful. It only comes up in certain specific situations.

This is *extremely* common.

Consider a Maybe/Option type.

```typescript
interface Option<T> {
  getOrElse(f: () => T): T;
}
```

A better type signature would be

```typescript
interface Option<T> {
  getOrElse<K super T>(f: () => K): K;
}
```

(Note: in this particular situation, Scala uses covariance. But Typescript doesn't have that.)

**Comment by mhegazy** [comment]
this should be covered by https://github.com/Microsoft/TypeScript/issues/11233 (and #4889).

**Comment by timsuchanek** [comment]
@GregRos this is an interesting idea and I also think the language would benefit from such a keyword.

As a workaround you can set an upper boundary in the param using a helper type:

```ts
type Full = {
  a: string
  b?: string
}

/**
 * Subset
 * @desc From `T` pick properties that exist in `U`
 */
type Subset<T, U> = { [key in keyof T]: key extends keyof U ? T[key] : never }

function acceptFullAndMore<T extends Full>(param: T): T {
  return param
}

acceptFullAndMore({a: '', c: 1}) // Compiler allows `c`, even if it's not part of `Full`

function acceptFull<T extends Full>(param: Subset<T, Full>): T {
  return param
}

acceptFull({a: '', c: 1}) // Compiler throws here 🙏
```

**Comment by laurensdijkstra** [comment]
@timsuchanek could your example be constrained even further to make the return type be of exactly the same type as the actual type passed to `param`? ie:
```
select<T extends Subset<T, Full>>(selector: T): T
```
or
```
select<T>(selector: Subset<T, Full>): Subset<T, Full>
```
but then actually working? 😛

**Comment by laurensdijkstra** [comment]
> this should be covered by #11233 (and #4889).

How could the previous example from @GregRos be achieved with the Partial<> type?
```
blah<Partial extended by Full>(partial : Partial) : Partial {
    return partial;
} //type information is preserved.
```
(where Partial is not actually the new Partial type from Typescript but his own type)

**Comment by studds** [comment]
I feel like I must be missing something obvious, but here's my use-case, which I just don't seem to be able to get working. I want to be able to pass in the subset of a generic with a key, and then use a property key & value in another function, but although the key is properly constrained, the property value is not:

```
type Full = {
  a: string
  b?: string
}

/**
 * Subset
 * @desc From `T` pick properties that exist in `U`
 */
type Subset<T, U> = { [key in keyof T]: key extends keyof U ? T[key] : never }

function acceptValueAndKey<K extends keyof Full>(key: K, value: Full[K]) {
  return {[key]: value}
}

function acceptFull<T extends Full>(param: Subset<T, Full>, key: keyof typeof param): T {
    acceptValueAndKey(param[key], key); // Compiler throws here - shouldn't - can't call in a way that would be invalid
  return param
}

acceptFull({a: ''}, 'a') // OK
acceptFull({a: '', b: ''}, 'b') // OK
acceptFull({a: '', b: ''}, 'b') // OK
acceptFull({a: '', c: ''}, 'a') // Compiler throws here (as it should)
```

Like @GregRos I'm surprised that I haven't found this exact issue - I might be searching for the wrong terms. If there's a more appropriate issue, please let me know.

**Comment by truemogician** [comment]
> @GregRos this is an interesting idea and I also think the language would benefit from such a keyword.
> 
> As a workaround you can set an upper boundary in the param using a helper type:
> 
> ```ts
> type Full = {
>   a: string
>   b?: string
> }
> 
> /**
>  * Subset
>  * @desc From `T` pick properties that exist in `U`
>  */
> type Subset<T, U> = { [key in keyof T]: key extends keyof U ? T[key] : never }
> 
> function acceptFullAndMore<T extends Full>(param: T): T {
>   return param
> }
> 
> acceptFullAndMore({a: '', c: 1}) // Compiler allows `c`, even if it's not part of `Full`
> 
> function acceptFull<T extends Full>(param: Subset<T, Full>): T {
>   return param
> }
> 
> acceptFull({a: '', c: 1}) // Compiler throws here 🙏
> ```

Such workaround requires a specific upper bound. However, in most cases the upper bound is quite universal, like `string`, `object` or simply `any`. We only want to filter out certain types without an exact super type.  
These scenarios seem to fit in the semantics of `Exclude<T, U>`. However, this utility type also requires the super type `T` to be *specific*, namely, union types. Universal types like `string` don't work as well, for instance, `Exclude<string, "foo">` results in `string`.   
This proposal is logically connected with #4196. With a negating type, assuming `Not<T>`, the upper-bound could be transformed into a lower-bound through the following utility types:
```ts
type UpperToLower<T extends object> = Partial<T> & Record<string & Not<keyof T>, never>;
```
So maybe this could be regarded as a sub-scenario to the #4196.

**Comment by graphemecluster** [comment]
Cross referencing #14520.
